src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -1173,6 +1140,150 @@ ChooseGenerator::ChooseGenerator(const Node& node, GeneratorArgs generatorArgs)
+}
+
+/**
+ * @private
+ * @param node
+ *   a node containing a date generator
+ * @return
+ *   a constant generator if given a constant/scalar.
+ */
+UniqueGenerator<int64_t> dateTimeGenerator(const Node& node, GeneratorArgs generatorArgs) {
+    // Set of parsers to look when we request an int parser
+    // see int64Generator
+    const static std::map<std::string, Parser<UniqueGenerator<bsoncxx::types::b_date>>> dateParsers{
+        {"^Now",
+         [](const Node& node, GeneratorArgs generatorArgs) {
+             return std::make_unique<NowGenerator>(node, generatorArgs);
+         }},
+        {"^RandomDate",
+         [](const Node& node, GeneratorArgs generatorArgs) {
+             return std::make_unique<RandomDateGenerator>(node, generatorArgs);
+         }},
+    };
+
+    if (auto parserPair = extractKnownParser(node, generatorArgs, dateParsers)) {
+        // known parser type
+        auto generator = parserPair->first(node[parserPair->second], generatorArgs);
+        auto datetime = generator->evaluate();
+        auto millis = datetime.to_int64();
+        return std::make_unique<ConstantAppender<int64_t>>(millis);
+    }
+    std::stringstream msg;
+    msg << "Malformed node for date. Not a datetime generator " << node;
+    BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax(msg.str()));
+}
+
+/**
+ * @param node
+ *   a node containing a date generator.
+ * generator
+ * @return
+ *   a constant generator if given a constant/scalar.
+ */
+UniqueGenerator<int64_t> dateGenerator(const Node& node,
+                                       GeneratorArgs generatorArgs,
+                                       const boost::posix_time::ptime& default_time) {
+    if (node) {
+        try {
+            auto generator = intGenerator(node, generatorArgs);
+            return generator;
+        } catch (const InvalidValueGeneratorSyntax& e) {
+            if (!isUnknownParser(e)) {
+                throw e;
+            }
+        } catch (const InvalidConversionException& e) {
+            BOOST_LOG_TRIVIAL(warning) << " InvalidConversionException: int";
+        }
+
+        try {
+            auto generator = doubleGenerator(node, generatorArgs);
+            return std::make_unique<ConstantAppender<int64_t>>((long long)generator->evaluate());
+        } catch (const InvalidValueGeneratorSyntax& e) {
+            if (!isUnknownParser(e)) {
+                throw e;
+            }
+        } catch (const InvalidConversionException& e) {
+        }
+
+        try {
+            auto generator = dateStringParserGenerator(node, generatorArgs);
+            return std::make_unique<ConstantAppender<int64_t>>((long long)generator->evaluate());
+        } catch (const InvalidValueGeneratorSyntax& e) {
+            if (!isUnknownParser(e)) {
+                throw e;
+            }
+        } catch (const InvalidConversionException& e) {
+        }
+
+        auto generator = dateTimeGenerator(node, generatorArgs);
```

Comment:
Should we order these by expected frequency? Like dateStringParser is probably most likely, then dateTimeGenerator. I'm thinking about all the try catches. I wonder if there is a way around it. It's a valid way to do it.